File: ACM International Collegiate Programming Contest, JUST Collegiate Programming Contest (2018)/k.cpp
#include <iostream>
#include <cstdlib>
#include <fstream>
#include <cstdio>
#include <climits>
#include <vector>
#include <map>
#include <list>
#include <deque>
#include <queue>
#include <stack>
#include <set>
#include <sstream>
#include <string>
#include <cstring>
#include <algorithm>
#include <bitset>
#include <cmath>
#include <utility>
#include <functional>
#include <cassert>

using namespace std;

string get_key(string& s) {
  int at = s.find('=');
  return s.substr(0, at);
}

string get_value(string& s) {
  int b = s.find('{');
  int e = s.find('}');
  return s.substr(b + 1, e - b - 1);
}

string format_author(string& s) {
  string ret = "";
  int par = 0;
  // cout << s << endl;
  for (int i = 0; i < s.length(); i++) {
    int j = i;
    while (j < s.length() && s[j] != ' ') {
      j++;
    }
    if (par % 2 == 0 && par > 0) {
      ret += ", ";
    }
    if (par % 2 == 0) {
      ret += s[i];
      ret += s[i + 1];
    } else {
      ret += ". ";
      ret += s[i];
    }
    // printf("%d %d\n", i, j);
    i = j;
    par++;
  }
  return ret;
}

int main() {
  ios_base::sync_with_stdio(false);

  int t;
  cin >> t;
  cin.ignore();
  while (t--) {
    int n;
    cin >> n;
    cin.ignore();
    while (n--) {
      map<string, string> m;
      string str;
      getline(cin, str);
      for (int i = 0; i < 8; i++) {
        getline(cin, str);
        string key = get_key(str);
        string value = get_value(str);
        m[key] = value;
      }
      getline(cin, str);

      printf("%s. %s. %s. %s;%s(%s):%s.\n",
        format_author(m["author"]).c_str(),
        m["title"].c_str(),
        m["journal"].c_str(),
        m["year"].c_str(),
        m["volume"].c_str(),
        m["number"].c_str(),
        m["pages"].c_str()
      );
    }
  }

  return 0;
}

File: ACM International Collegiate Programming Contest, JUST Collegiate Programming Contest (2018)/k.java
import java.io.*;
import java.util.*;
import java.lang.*;
import java.math.BigInteger;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

/**
 * Built using CHelper plug-in
 * Actual solution is at the top
 */

public class k {
  public static void main(String[] args) {
    Scanner in = new Scanner(System.in);
    OutputStream out = System.out;
    Task solver = new Task();
    solver.solve(1, in, out);
  }
}

class Task {
  String getKey(String s) {
    int at = s.indexOf('=');
    return s.substring(0, at);
  }
  String getValue(String s) {
    Pattern p = Pattern.compile("\\{(.*?)\\}");
    Matcher m = p.matcher(s);

    while (m.find()) {
      return m.group(1);
    }

    return "";
  }
  String formatAuthor(String s) {
    String names[] = s.split(",");
    StringBuilder ret = new StringBuilder();
    for (int i = 0; i < names.length; i++) {
      if (i > 0) {
        ret.append(", ");
      }
      String name = names[i].trim();
      String sname[] = name.split(" ");
      ret.append(sname[0].charAt(0));
      ret.append(sname[0].charAt(1));
      ret.append(". ");
      ret.append(sname[1].charAt(0));
    }
    return ret.toString();
  }

  public void solve(int testNumber, Scanner in, OutputStream out) {
    int t = Integer.parseInt(in.nextLine());
    while (t-- > 0) {
      int n = Integer.parseInt(in.nextLine());
      while (n-- > 0) {
        in.nextLine();
        Map map = new HashMap();
        for (int i = 0; i < 8; i++) {
          String str = in.nextLine();
          String key = getKey(str);
          String value = getValue(str);
          map.put(key, value);
        }
        in.nextLine();

        System.out.printf("%s. %s. %s. %s;%s(%s):%s.\n",
          formatAuthor((String)map.get("author")),
          map.get("title"),
          map.get("journal"),
          map.get("year"),
          map.get("volume"),
          map.get("number"),
          map.get("pages")
        );
      }
    }
  }
}
